### PR TITLE
Remove all sorts of deprecated stuff in the Database namespace

### PR DESCRIPTION
**Pull request by Toflar**

### CODE REVIEW

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php [hunk 1]
```diff
@@ -228,16 +220,7 @@ public function limit($intRows, $intOffset=0)
...
-
-		if (\count($arrParams) === 1 && \is_array($arrParams[0]))
-		{
-			trigger_deprecation('contao/core-bundle', '4.13', 'Using "%s()" with an array parameter has been deprecated and will no longer work in Contao 5.0. Use argument unpacking via ... instead."', __METHOD__);
-
-			$arrParams = array_values($arrParams[0]);
-		}
-
-		return $this->query('', array_merge($this->arrSetParams, $arrParams));
+		return $this->query('', array_merge($this->arrSetParams, \func_get_args()));
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/Database/Statement.php:228]: we could use `...$args` in the method name and here?

Comment (Toflar) [reply core-bundle/src/Resources/contao/library/Contao/Database/Statement.php:228]: We could but I'm removing deprecated stuff here, I'm not using new language constructs.

**Review by aschempp — DISMISSED** [review]

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php [hunk 2]
```diff
@@ -141,13 +140,6 @@ public function prepare($strQuery)
 	 */
 	public function set($arrParams)
 	{
-		if (substr_count($this->strQuery, '%s') !== 1 || !\in_array(strtoupper(substr($this->strQuery, 0, 6)), array('INSERT', 'UPDATE'), true))
-		{
-			trigger_deprecation('contao/core-bundle', '4.13', 'Using "%s()" is only supported for INSERT and UPDATE queries with the "%%s" placeholder. This will throw an exception in Contao 5.0.', __METHOD__);
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Database/Statement.php:146]: Where is the exception thrown?

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php [hunk 3]
```diff
@@ -228,16 +220,7 @@ public function limit($intRows, $intOffset=0)
 	 */
 	public function execute()
 	{
-		$arrParams = \func_get_args();
-
-		if (\count($arrParams) === 1 && \is_array($arrParams[0]))
-		{
-			trigger_deprecation('contao/core-bundle', '4.13', 'Using "%s()" with an array parameter has been deprecated and will no longer work in Contao 5.0. Use argument unpacking via ... instead."', __METHOD__);
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Database/Statement.php:235]: Did you check that we don‘t accidentally do this?

Comment (Toflar) [reply core-bundle/src/Resources/contao/library/Contao/Database/Statement.php:235]: I expect us not to if it's deprecated.

File: core-bundle/src/Resources/contao/library/Contao/Database/Statement.php (hunk 2 above)
Comment (Toflar) [reply core-bundle/src/Resources/contao/library/Contao/Database/Statement.php:146]: I guess Doctrine will? @ausi ?

Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/Database/Statement.php:146]: The `trigger_deprecation()` has to be replaced with a `throw` statement here.

File: core-bundle/tests/Contao/Database/StatementTest.php [hunk 4]
```diff
@@ -1,333 +0,0 @@
...
-use Contao\System;
-use Doctrine\DBAL\Connection;
-use Doctrine\DBAL\Driver\Mysqli\Exception\ConnectionError;
-use Doctrine\DBAL\Exception\DriverException;
-use Doctrine\DBAL\Platforms\AbstractPlatform;
-use Doctrine\DBAL\Result;
-use Symfony\Bridge\PhpUnit\ExpectDeprecationTrait;
-use Symfony\Component\DependencyInjection\Container;
-
-class StatementTest extends TestCase
```
Comment (ausi) [inline core-bundle/tests/Contao/Database/StatementTest.php]: This should not be removed here IMO, there are tests for the correct behavior in here too (not only tests for deprecations).

File: core-bundle/tests/Functional/SearchTest.php [hunk 5]
```diff
@@ -1,101 +0,0 @@
...
-    {
-        unset($GLOBALS['TL_LANGUAGE']);
-
-        parent::tearDown();
-    }
-
-    /**
-     * @group legacy
-     */
-    public function testSearchFor(): void
```
Comment (ausi) [inline core-bundle/tests/Functional/SearchTest.php:45]: Should be kept and rewired to `Search::query()`

File: core-bundle/tests/Contao/Database/StatementTest.php (hunk 4 above)
Comment (Toflar) [reply core-bundle/tests/Contao/Database/StatementTest.php]: Which ones? They all contain `$this->expectDeprecation()`.

File: core-bundle/tests/Functional/SearchTest.php (hunk 5 above)
Comment (Toflar) [reply core-bundle/tests/Functional/SearchTest.php:45]: Why are these tests so strange? We should be able to delete all methods of deprecated stuff and new ones should be tested separately, no?

File: core-bundle/tests/Contao/Database/StatementTest.php [hunk 6]
```diff
@@ -1,333 +0,0 @@
...
-     */
-    public function testSetTriggersDeprecation(string $query): void
-    {
-        $statement = new Statement($this->createMock(Connection::class));
-
-        if ($query) {
-            $statement->prepare($query);
-        }
-
-        $this->expectDeprecation('%sUsing "%s()" is only supported for INSERT and UPDATE queries with the "%%s" placeholder%s');
```
Comment (ausi) [inline core-bundle/tests/Contao/Database/StatementTest.php:49]: Should be replaced with expecting an exception.

File: core-bundle/tests/Contao/Database/StatementTest.php [hunk 7]
```diff
@@ -1,333 +0,0 @@
...
-        $statement = (new Statement($connection));
-        $statement->prepare($query);
-
-        if ($set) {
-            $statement->set($set);
-        }
-
-        $this->expectDeprecation('%sUsing "%s" with an array parameter has been deprecated%s');
-
-        $statement->execute($params ?? []);
```
Comment (ausi) [inline core-bundle/tests/Contao/Database/StatementTest.php:211]: Everything except these last lines from this test should stay IMO.

**Review by ausi — CHANGES_REQUESTED** [review]

File: core-bundle/tests/Functional/SearchTest.php (hunk 5 above)
Comment (ausi) [reply core-bundle/tests/Functional/SearchTest.php:45]: Probably “should” yes. But AFAIK the legacy code is not 100% covered by tests.
This class is the only search test we have AFAIK, we should not remove it.

**Comment by Toflar** [comment]
Sorry, I won't be able to finish this PR because tests are not clearly separated and I don't know how they have to be adjusted.

**Comment by ausi** [comment]
Should I take over this one?

**Comment by Toflar** [comment]
Sure, you can either PR against my PR or whatever is easier for you.

**Review by ausi — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.

**Pull request merged**